### PR TITLE
PersistentVolumeSet API & Client

### PR DESCRIPTION
**Pull request by markturansky**
See https://github.com/GoogleCloudPlatform/kubernetes/pull/6773

### CODE REVIEW

**Comment by markturansky** [comment]
@jsafrane @swagiaal  first review?

**Comment by markturansky** [comment]
Rebased.

File: docs/design/persistent-volume-provisioning.md [hunk 1]
```diff
@@ -0,0 +1,159 @@
...
+</strong>
+--
+
+<!-- END STRIP_FOR_RELEASE -->
+
+<!-- END MUNGE: UNVERSIONED_WARNING -->
+
+# Persistent Volume Provisioning
+
+This document proposes a model for dynamically provisioning
```
Comment (swagiaal) [inline docs/design/persistent-volume-provisioning.md]: "dynamically provisioning"... persistent volumes ?

File: docs/design/persistent-volume-provisioning.md [hunk 2]
```diff
@@ -0,0 +1,159 @@
...
+
+### tl;dr
+
+One new API kind:
+
+A `PersistentVolumeController` (PVC) is a storage resource provisioned by an administrator.  PVCs do not have a namespace.  Just as a `ReplicationController` maintains a number of replicas of a pod, a `PersistentVolumeController` maintains a minimum number of replicas of a `PersistentVolume`.  A PVC creates new volumes from a template up to a maximum replica count in increments of the minimum replica count.  A well-provisioned cluster will have many PVCs.
```
Comment (swagiaal) [inline docs/design/persistent-volume-provisioning.md]: "A PVC creates new volumes from a template up to a maximum replica count in increments of the minimum replica count"
Hmm... Does that mean that if minimumReplicas is 3 that volumes will be created 3 volumes at a time ? Why is that desired ?

File: docs/design/persistent-volume-provisioning.md [hunk 3]
```diff
@@ -0,0 +1,159 @@
...
+
+// PersistentVolumeControllerSpec is the specification of a PV controller.
+// The controller creates volumes exactly like the Template until the unbound count equals the replica count
+type PersistentVolumeControllerSpec struct {
+	// MinimumReplicas is the minimum number of unbound persistent volumes of this type desired in the system maintain
+	MinimumReplicas int `json:"minimumReplicas"`
+
+	// MaximumReplicas is the maximum total number of persistent volumes desired in the system
+	MaximumReplicas int `json:"maximumReplicas"`
+
```
Comment (swagiaal) [inline docs/design/persistent-volume-provisioning.md]: Hmm for what it is worth I found the above terms confusing at first pass. How about s/minimum/minimumUnbound ?
or just "replicas" to use the same terminology as replication controllers. maximumReplicas refers to the total of bound and unbound replicas. Correct ?

File: docs/design/persistent-volume-provisioning.md [hunk 4]
```diff
@@ -0,0 +1,159 @@
...
+	// Metadata of the PVs created from this template.
+	ObjectMeta `json:"metadata,omitempty"`
+
+	// Spec defines a persistent volume.
+	Spec PersistentVolumeSpec `json:"spec,omitempty"`
+}
+
+```
+
+Dynamic pools of storage (e.g, GCE) will maintain `MinimumReplicas` of a `PersistentVolume` at a given time where each PVs status phase is ` Available`, up to `MaximumReplicas`.  Storage levels will increase by one until the minimum or maximum levels are reached.  Recycling a PV means the current number of available volumes might exceed the minimum replica count.  Plugins implementing the Deleter interface will delete PVs where the `PersistentVolumeReclaimPolicy` is set to Delete.  The default reclaim policy is Retain.
```
Comment (swagiaal) [inline docs/design/persistent-volume-provisioning.md]: > Storage levels will increase by one until the minimum or maximum levels are reached
Ah! this seems to contradict the statement I commented on above but I think this is the correct one :)

Comment (markturansky) [inline docs/design/persistent-volume-provisioning.md]: Yeah, and I missed the other reference.  I changed this doc in the proposal PR to make this clearer.

https://github.com/kubernetes/kubernetes/pull/6773

File: docs/design/persistent-volume-provisioning.md (hunk 1 above)
Comment (markturansky) [inline docs/design/persistent-volume-provisioning.md]: Rendering issue?  There's a link there to the Persistent Volumes doc.  It's not supposed to just cut off...

Comment (swagiaal) [inline docs/design/persistent-volume-provisioning.md]: hmm.. I don't see it. I looked at the diff and the rendered file. The sentence ends with "dynamically provisioning"
Maybe I am looking at the wrong version

Comment (markturansky) [inline docs/design/persistent-volume-provisioning.md]: Yeah, it's in the PR for the proposal.  I haven't been messing with the *.md doc in this PR.  should probably just remove it entirely from this PR.

**Comment by markturansky** [comment]
I'm closing this but will retain the branch in case we want this API object in the future.

**Pull request closed without merging**